auction-house/program/src/lib.rs

```diff
@@ -965,6 +966,96 @@ pub mod auction_house {
         Ok(())
     }
 
+    pub fn print_receipt<'info>(
+        ctx: Context<'_, '_, '_, 'info, PrintReceipt<'info>>,
+        receipt_bump: u8,
+        buyer_price: u64,
+        token_size: u64,
+    ) -> ProgramResult {
+        let trade_state = &ctx.accounts.trade_state;
+        let receipt = &mut ctx.accounts.receipt;
+        let auction_house = &ctx.accounts.auction_house;
+        let token_account = &ctx.accounts.token_account;
+        let bookkeeper = &ctx.accounts.bookkeeper;
+        let wallet = &ctx.accounts.wallet;
+
+        let trade_state_bump = trade_state.to_account_info().data.borrow()[0];
+
+        receipt.trade_state = receipt.trade_state;
+        receipt.auction_house = auction_house.key();
+        receipt.token_account = token_account.key();
+        receipt.bookkeeper = bookkeeper.key();
+        receipt.wallet = wallet.key();
+        receipt.buyer_price = buyer_price;
+        receipt.token_size = token_size;
+        receipt.bump = receipt_bump;
+
+        let ts_seeds = [
+            PREFIX.as_bytes(),
+            receipt.wallet.as_ref(),
+            receipt.auction_house.as_ref(),
+            receipt.token_account.as_ref(),
+            auction_house.treasury_mint.as_ref(),
+            token_account.mint.as_ref(),
+            &receipt.buyer_price.to_le_bytes(),
+            &receipt.token_size.to_le_bytes(),
+            &[trade_state_bump],
+        ];
+
+        assert_derivation(
+            &id(),
+            &trade_state.to_account_info(),
+            &ts_seeds,
+        )?;
```

Comment:
This should be ensuring that receipts are only saved when the contents of the receipt equate to the seed of the tradestate ie only make receipts that match tradestates.